framework/include/dirackernels/DiracKernelBase.h

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * DiracKernelBase is the base class for all DiracKernel type classes.
+ */
+class DiracKernelBase : public ResidualObject,
+                        public CoupleableMooseVariableDependencyIntermediateInterface,
+                        public MaterialPropertyInterface,
+                        protected GeometricSearchInterface,
+                        public BlockRestrictable
+{
+public:
+  static InputParameters validParams();
+
+  DiracKernelBase(const InputParameters & parameters);
+
+  /**
+   * Computes the residual for the current element.
+   */
+  virtual void computeResidual() override = 0;
+
+  /**
+   * Computes the jacobian for the current element.
+   */
+  virtual void computeJacobian() override = 0;
+
+  /**
+   * This gets called by computeOffDiagJacobian() at each quadrature point.
+   */
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
+
+  /**
+   * Computes the off-diagonal Jacobian for variable jvar.
+   */
+  virtual void computeOffDiagJacobian(unsigned int jvar) override = 0;
+
+  /**
+   * This is where the DiracKernel should call addPoint() for each point it needs to have a
+   * value distributed at.
+   */
+  virtual void addPoints() = 0;
+
+  /**
+   * Whether or not this DiracKernel has something to distribute on this element.
+   */
+  bool hasPointsOnElem(const Elem * elem);
+
+  /**
+   * Whether or not this DiracKernel has something to distribute at this Point.
+   */
+  bool isActiveAtPoint(const Elem * elem, const Point & p);
+
+  /**
+   * Remove all of the current points and elements.
+   */
+  void clearPoints();
+
+  /**
+   * Clear point cache when the mesh changes, so that element
+   * coarsening, element deletion, and distributed mesh repartitioning
+   * don't leave this with an invalid cache.
+   */
+  virtual void meshChanged() override;
+
+protected:
+  /**
+   * This is the virtual that derived classes should override for computing the residual.
+   */
+  virtual Real computeQpResidual() = 0;
+
+  /**
+   * This is the virtual that derived classes should override for computing the Jacobian.
+   */
+  virtual Real computeQpJacobian();
+
+  /**
+   * Add the physical x,y,z point located in the element "elem" to the list of points
+   * this DiracKernel will be asked to evaluate a value at.
+   */
+  void addPoint(const Elem * elem, Point p, unsigned id = libMesh::invalid_uint);
+
+  /**
+   * This is a highly inefficient way to add a point where this DiracKernel needs to be
+   * evaluated.
+   *
+   * This spawns a search for the element containing that point!
+   */
+  const Elem * addPoint(Point p, unsigned id = libMesh::invalid_uint);
+
+  /**
+   * Returns the user-assigned ID of the current Dirac point if it
+   * exits, and libMesh::invalid_uint otherwise.  Can be used e.g. in
+   * the computeQpResidual() function to determine the cached ID of
+   * the current point, in case this information is relevant.
+   */
+  virtual unsigned currentPointCachedID() = 0;
```

Comment:
Is there any reason why this method can't be implemented in the base class, in which case we don't need to make this function `virtual`?

Reply:
This function needs `_current_elem`. `_current_elem` needs to be created in the DiracKernelTempl class (as far as I understand). It didn't need the `virtual`,though, so I changed that.